### DISCUSSION TITLE
build with grub from backports

### DISCUSSION THREAD

**Discussion by marwenius** [SpiralLinux/SpiralLinux-project#442] [Q&A] [answered]
I would like to use Spirallinux on a C13 Yoga with coreboot (more information: [Chrultrabook](https://docs.chrultrabook.com/)). Unfortunately this does not work with version 2.06 of grub, but with version 2.12 from the backports (tested with a live image of the testing branch which uses the same version).

If I now want to create my own live image, do I only have to change the `99backports` file accordingly or do I need to make other adjustments?

What I would add:
```
Package: grub*
Pin: release a=stable-backports
Pin-Priority: 600
```

Thanks in advance!

**Comment by geckolinux** [comment]
Hi there, yes, that looks right. 
In my research I read something about how package `name*` wildcards aren't supported, but I've also seen examples like yours, so please let us know if it works or not.

**Reply by marwenius** [reply]
It runs almost to the end, but then this error occurs:
```
[2024-06-22 17:23:02] lb binary_grub-efi 
P: Begin preparing Grub based EFI support...
Reading package lists...
Building dependency tree...
Reading state information...
shim-signed is already the newest version (1.39+15.7-1).
grub-efi-amd64-signed is already the newest version (1+2.12+1~bpo12+1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
P: UEFI Secure Boot support enabled.
Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 grub-efi-ia32-bin : Breaks: grub-efi-ia32-signed:i386 (< 1+2.12~rc1) but 1+2.06+13+deb12u1 is to be installed
E: Unable to correct problems, you have held broken packages.
E: An unexpected failure occurred, exiting...
P: Begin unmounting filesystems...
P: Saving caches...
Reading package lists...
Building dependency tree...
Reading state information...
mv: der Aufruf von stat für 'live-image-amd64.hybrid.iso' ist nicht möglich: Datei oder Verzeichnis nicht gefunden
```

I will try another run later with adjusted values in `99backports`.

**Reply by marwenius** [reply]
I have given up this approach for now. Either the build process runs through, but the version of grub does not change or the build process aborts near the end due to missing dependencies.

That's why I've now solved it differently. I use two USB sticks. Debian testing is on one and Spirallinux on the other.

With the grub shell of Debian testing I manually load the necessary files from Spirallinux to be able to boot the live system:
```
set root=(cd0)
linux /live/vmlinuz boot=live root=/dev/sr1
initrd /live/initrd.img
boot
```
(maybe it could be even more minimalist/simpler, but this worked for me)

Depending on the system, you may have to use `sr0` or `(cd1)` instead of `sr1` or `(cd0)` (or other values).

In any case, I was able to start the live system and install Spirallinux. Then only the grub version 2.06 is installed, but the system can be started with the same trick (albeit somewhat more complicated/extensive) and as soon as you are in, you install `grub-efi` from the backports and run `update-grub` in addition to `grub-install`.

If you have any ideas on how the building process could work, I can try it out. Unfortunately, I don't have the knowledge to find a solution on my own (and after several attempts, which always take at least a quarter of an hour to build, I'm also a bit tired).

**Reply by marwenius** [reply, marked as answer]
What an irony. I'm about to give up and think to myself "one last try" (*cough*) and what can I say: it worked after all.

In `config/includes.chroot/etc/apt/preferences.d/99backports` I have added:
```
Package: grub-common grub2-common grub-pc-bin grub-efi grub-efi-amd64 grub-efi-amd64-bin grub-efi-amd64-signed grub-efi-amd64-signed-template grub-efi-ia32-bin
Pin: release a=stable-backports
Pin-Priority: 600
```
And in `config/package-lists/spirallinux-Cinnamon-packages.list.chroot` I have added the following lines:
```
grub-efi-amd64
grub-efi
```
And I had to comment out the following lines:
```
#grub-efi-ia32-signed:i386
#grub-efi-ia32-signed-template:i386
```

I also had to manually download some packages into the `config/packages` directory:
```
sudo apt download -t bookworm-backports linux-image-amd64 linux-headers-amd64 grub-common grub2-common grub-pc-bin grub-efi grub-efi-amd64 grub-efi-amd64-bin grub-efi-amd64-signed grub-efi-amd64-signed-template grub-efi-ia32-bin
```
Now the live system has the required version of grub and this is also installed on the computer during the installation.

Unexpected side effects:

The live system is only delivered with the backport kernel, just like the installation.

And the `script firstboot-update-initramfs.service` [which is supposed to run after the first reboot](https://github.com/SpiralLinux/SpiralLinux-project/issues/328#issuecomment-1797627824) is no longer executed. Since I know the error, I can fix it manually.

Do you have an idea why this happens and how it can be fixed?

But as long as it's just this one thing, I can live with it. Or are there other things I should look out for?
